src/compiler/checker.ts

```diff
@@ -11534,26 +11632,26 @@ namespace ts {
                 let result = root.instantiations!.get(id);
                 if (!result) {
                     const newMapper = createTypeMapper(root.outerTypeParameters, typeArguments);
-                    result = instantiateConditionalType(root, newMapper);
+                    result = yield* instantiateConditionalType(root, newMapper);
                     root.instantiations!.set(id, result);
                 }
                 return result;
             }
             return type;
         }
 
-        function instantiateConditionalType(root: ConditionalRoot, mapper: TypeMapper): Type {
+        function* instantiateConditionalType(root: ConditionalRoot, mapper: TypeMapper): Generator<[Type, TypeMapper], Type, Type> {
             // Check if we have a conditional type where the check type is a naked type parameter. If so,
             // the conditional type is distributive over union types and when T is instantiated to a union
             // type A | B, we produce (A extends U ? X : Y) | (B extends U ? X : Y).
             if (root.isDistributive) {
                 const checkType = <TypeParameter>root.checkType;
                 const instantiatedType = mapper(checkType);
                 if (checkType !== instantiatedType && instantiatedType.flags & (TypeFlags.Union | TypeFlags.Never)) {
-                    return mapType(instantiatedType, t => getConditionalType(root, createReplacementMapper(checkType, t, mapper)));
+                    return yield* mapTypeStaged(instantiatedType, function* (t) { return yield* getConditionalType(root, createReplacementMapper(checkType, t, mapper)); });

```

Comment:
FYI, lines like this one are why generator arrow functions might actually be nice. Before writing this I couldn't see why I'd want them, but after going through the motions of a trampoline rewrite, being forced to replace an `=>` with a `function*` expression just to `yield*` within it... `=*>` or `=>*` would be nice (esp. considering `async` arrows can already compose like this).

Reply:
It’s tricky though; typically a generator will yield more than one thing so you will almost always need to add braces for a multiline arrow.  At that point the only thing you gain from the arrow syntax is the automatic `this` binding, which I haven’t personally found a need for.

Even in single-line cases (`yield*`), returning the result of the yield is unlikely to be what you want *in general*.  I can definitely understand why there’s no support for it.

Reply:
Erm, the use case is simply rebinding or reprocessing arguments, IMO. Just like how `x => call(x + 1)` is useful, `x =>* yield* gen(x + 1)` would be useful.

Reply:
I know, I just meant that if that `yield*` returns anything other than `undefined` (is that possible?), that will also be returned by the arrow generator as its `done: true` case, which I think might violate the principle of least astonishment in general.  It’s not perfectly symmetrical with the regular function case, where passing through the return value of the chained call *is* generally what you want.

Sorry, I like to think in very abstract general cases, sometimes to my own detriment. 😛

Reply:
`yield*` returns the `return` value of the `yield*`'d generator. So if I have
```ts
function* foo() {
   return 2;
}
```
then `yield* foo()` is `2`. In this code that matters a _lot_ - the return type of the generator is the actual return type of the operation (while the yield type is the arguments to instantiate).